### PR TITLE
[r] Support timestamps (in READ mode)

### PR DESCRIPTION
**Pull request by mlin**
#1347 

All create/open factory methods take a new argument `tiledb_timestamp` which can be set to a `POSIXct` datetime. This is passed down into TileDB-R appropriately, including propagation from collection objects to their members. 

This PR only supports/tests the timestamps for READ mode. We can follow up to add WRITE mode support. Maybe quickly, or maybe not (since the API wouldn't change, and setting write timestamps isn't a high-priority need)

### CODE REVIEW

File: apis/r/R/Factory.R [hunk 1]
```diff
@@ -6,10 +6,11 @@
 #' @param index_column_names Index column names passed on to DataFrame$create()
 #' @param platform_config Optional platform configuration
 #' @param tiledbsoma_ctx Optional SOMATileDBContext
+#' @param tiledb_timestamp Optional POSIXct for TileDB timestamp
```
Comment (eddelbuettel) [inline apis/r/R/Factory.R:9]: Maybe write as 'Optional Datetime (i.e. POSIXct) variable with TileDB timestamp.  Not everybody will be familiar with term POSIXct.

**Comment by eddelbuettel** [comment]
I built on top of your commit in two further commit.  As this branch came up 'detached head' for me, I created a new one, cherry-picked this and added.  Feel free to reverse cherry pick.  Changes are in https://github.com/single-cell-data/TileDB-SOMA/tree/de/sc-28834/timestamped_open

**Comment by eddelbuettel** [comment]
So the failng test fails ... because of something that I need to fix in tiledb-r (i.e.: pass through (fake) 'written at' timestamps).  If we rewrite the test in the way the tiledb-r does it (i.e.: not when written, adjust read time) it passes for me (but I have some more changes here in branch I may want to clean up before commiting).  For me, this now passes:

```r
dir.create(uri <- tempfile())

## create 2x2 array and write 1 into top-left entry
t1 <- Sys.time()
snda <- SOMASparseNDArrayCreate(uri=uri, type=arrow::int16(), shape=c(2,2)) #, tiledb_timestamp=ts(10))
snda$write(Matrix::sparseMatrix(i = 1, j = 1, x = 1, dims = c(2, 2)))
snda$close()
Sys.sleep(1.0)
t2 <- Sys.time()
## write 1 into bottom-right entry
snda <- SOMASparseNDArrayOpen(uri=uri, mode="WRITE") #, tiledb_timestamp=ts(20))
snda$write(Matrix::sparseMatrix(i = 2, j = 2, x = 1, dims = c(2, 2)))
snda$close()

## read with no timestamp args and see both writes
snda <- SOMASparseNDArrayOpen(uri=uri)
mat <- snda$read()$sparse_matrix()$concat()
expect_equal(sum(mat), 2)
snda$close()

## read in between timestamps and see only the first write
snda <- SOMASparseNDArrayOpen(uri=uri, tiledb_timestamp = t1 + 0.5*as.numeric(t2-t1))
mat <- snda$read()$sparse_matrix()$concat()
expect_equal(sum(mat), 1)
snda$close()
```

File: apis/r/R/TileDBGroup.R [hunk 2]
```diff
@@ -53,24 +53,28 @@ TileDBGroup <- R6::R6Class(
...
+        if (mode == "WRITE") {
+          warning("Setting tiledb_timestamp is not recommended in WRITE mode!")
+        }
         spdl::debug("Opening {} '{}' in {} mode at {}",
                     self$class(), self$uri, mode, private$.group_open_timestamp)
         ## The Group API does not expose a timestamp setter so we have to go via the config
         cfg <- tiledb::config(private$.tiledb_ctx)
-        cfg["sm.group.timestamp_end"] <- private$.group_open_timestamp
+        cfg["sm.group.timestamp_end"] <- format(round(as.numeric(private$.group_open_timestamp) * 1000),
+                                                scientific = FALSE, big.mark = "")
```
Comment (eddelbuettel) [inline apis/r/R/TileDBGroup.R:77]: Thanks.  That is a good careful use of the required 'to character' conversion.

File: apis/r/R/TileDBArray.R [hunk 3]
```diff
@@ -24,12 +24,15 @@ TileDBArray <- R6::R6Class(
       private$.mode <- mode
       if (is.null(private$tiledb_timestamp)) {
         spdl::debug("[TileDBArray::open] Opening {} '{}' in {} mode", self$class(), self$uri, mode)
-        tiledb::tiledb_array_open(self$object, type = mode)
+        private$.tiledb_array <- tiledb::tiledb_array_open(self$object, type = mode)
```
Comment (eddelbuettel) [inline apis/r/R/TileDBArray.R:27]: I am a bit puzzled that we got by so far without an explicit copy and that this requires it now.

File: apis/r/R/TileDBArray.R [hunk 4]
```diff
@@ -24,12 +24,15 @@ TileDBArray <- R6::R6Class(
...
       } else {
+        if (mode == "WRITE") {
+          warning("Setting tiledb_timestamp is not recommended in WRITE mode!")
+        }
```
Comment (eddelbuettel) [inline apis/r/R/TileDBArray.R:31]: Is that a general policy, i.e. does Python also warn?

File: apis/r/R/TileDBGroup.R [hunk 5]
```diff
@@ -53,24 +53,28 @@ TileDBGroup <- R6::R6Class(
...
         spdl::debug("Opening {} '{}' in {} mode", self$class(), self$uri, mode)
         private$.tiledb_group <- tiledb::tiledb_group(
           self$uri,
           type = mode,
           ctx = private$.tiledb_ctx
         )
       } else {
+        if (mode == "WRITE") {
+          warning("Setting tiledb_timestamp is not recommended in WRITE mode!")
+        }
```
Comment (eddelbuettel) [inline apis/r/R/TileDBGroup.R:71]: As above.

File: apis/r/R/TileDBGroup.R [hunk 6]
```diff
@@ -391,6 +395,7 @@ TileDBGroup <- R6::R6Class(
       # we must open a temporary handle for read, to fill the cache.
       group_handle <- private$.tiledb_group
       if (private$.mode == "WRITE") {
+        stopifnot("FIXME" = is.null(private$.group_open_timestamp))
```
Comment (eddelbuettel) [inline apis/r/R/TileDBGroup.R:398]: Is this supposed to remain?

**Review by eddelbuettel — COMMENTED** [review]
This looks good but I left a few questions.

**Comment by mlin** [comment]
@eddelbuettel Seems like there's something weird when the result set is empty for `SOMASparseNDArray$read()$sparse_matrix()$concat()` which (I infer from your last commit message) neither of us see locally. 

```
  ── Error ('test-SOMACollection.R:99:3'): SOMACollection timestamped ops ────────
  Error in `Matrix::sparseMatrix(i = soma_dim_0_one_based, j = soma_dim_1_one_based, 
      x = soma_data, dims = shape, repr = repr)`: all(dims >= dims.min) is not TRUE
  Backtrace:
      ▆
   1. └─collection$get("A")$read()$sparse_matrix()$concat() at test-SOMACollection.R:99:2
   2.   └─self$read_next()
   3.     └─private$soma_reader_transform(rl)
   4.       └─tiledbsoma:::arrow_table_to_sparse(...)
   5.         └─Matrix::sparseMatrix(...)
   6.           └─base::stopifnot(all(dims >= dims.min))
   ```

**Comment by eddelbuettel** [comment]
Yes -- I even rotated over `tiledb` version (CRAN, mine) for no change.  Fiddling with the time time delta did not help so I would suggest to condition this away if, say, `Sys.getenv("CI", "")` is not empty.  Thoughts?

File: apis/r/R/TileDBArray.R (hunk 3 above)
Comment (mlin) [reply apis/r/R/TileDBArray.R:27]: The problem I'd hit without this reassignment seemed to be specific to setting WRITE timestamp, which our tests (in this PR) are not doing. But I added this based on the discussion on the TileDB-R thread and to be consistent with TileDBGroup.R (still not sure if this change to TileDBArray.R had gotten lost in rebase or smth)

**Comment by mlin** [comment]
@eddelbuettel Worked around by making both test result sets nonempty. I think that's most probably unrelated to the timestamps -- I'll file a separate issue to follow up

File: apis/r/R/TileDBGroup.R (hunk 6 above)
Comment (mlin) [reply apis/r/R/TileDBGroup.R:399]: @eddelbuettel I changed the WRITE mode warnings to stops which makes this "FIXME" unreachable (also added comments to this effect).

The python implementation doesn't raise a warning but there are [docstring allusions](https://github.com/single-cell-data/TileDB-SOMA/blob/58f528cb602e840b34696b93958f0731e998e0f2/apis/python/src/tiledbsoma/options/_soma_tiledb_context.py#L56-L58) to some of the pitfalls one can fall into by setting the write timestamp for multiple operations. I was personally disturbed enough by this (#919) to resolve to (personally) avoid the feature from then on. Also I think our main use case is that there should *exist some timestamps* where user will get a consistent snapshot, not that there should *exist no timestamps* at which a user would get an inconsistent snapshot. Reasonable people can and did disagree about all this, of course.

tl;dr I think we're fine leaving this unsupported for now, if we do want to work on it (in the next sprint or perhaps much later) then it wouldn't change the API, so needn't block 1.0rc either.

Comment (eddelbuettel) [reply apis/r/R/TileDBGroup.R:399]: Thanks for the explanation.  Now, without going into specifics of this question, I am generally a bigger fan of "document and explain and lead to _preferred_ interfaces" (as we eg do for open and create factories utilising the available-but-discouraged R6 methods by also requiring the 'handshake') and hence not as much a fan of hard locks such as this one.  I will need to ponder this a little more, and we may want to ascertain we get additional reviews and buy in for the PR.

File: apis/r/tests/testthat/test-SOMASparseNDArray.R [hunk 7]
```diff
@@ -309,3 +308,34 @@ test_that("platform_config defaults", {
...
 })
+
+test_that("SOMASparseNDArray timestamped ops", {
+  uri <- withr::local_tempdir("soma-sparse-nd-array-timestamps")
+
+  ts <- function(s) as.POSIXct(s, origin="1970-01-01", tz="UTC")
+
+  # t=10: create 2x2 array and write 1 into top-left entry
+  ## NB: timestamp-on-write do not currently go through on write via tiledb-r
+  ##     this needs to be addressed first in tiledb-r, this package can then catch up
```
Comment (johnkerl) [inline apis/r/tests/testthat/test-SOMASparseNDArray.R:319]: Please include, as a code comment here, a link to the issue which tracks this to-do

**Review by johnkerl — APPROVED** [review]

File: apis/r/tests/testthat/test-SOMASparseNDArray.R (hunk 7 above)
Comment (eddelbuettel) [reply apis/r/tests/testthat/test-SOMASparseNDArray.R:319]: This actually is now in tiledb-r albeit not yet in the CRAN version.

Comment (mlin) [reply apis/r/tests/testthat/test-SOMASparseNDArray.R]: Thanks @johnkerl I cleaned up these comments for now but tracking it in broken-out #1450
